### DISCUSSION TITLE
What does it mean for a Python package to be pure?

### DISCUSSION THREAD

**Discussion by mathewdblewis** [pyodide/pyodide#2056] [Q&A] [answered]
In the readme on the pyodide github here https://github.com/pyodide/pyodide/#what-is-pyodide it states the following:
"Any pure Python package with a wheel available on PyPi is supported"

What exactly is meant by the word 'pure' in this context. For example, I have tried using both PyPDF2 and PyPDF4:
https://pypi.org/project/PyPDF2/
https://pypi.org/project/PyPDF4/

Both of which make the following claim in their project description:
"A Pure-Python library built as a PDF toolkit"

However when I attempt to import either of these libraries into pyodide, I get the following error:
ValueError: Couldn't find a pure Python 3 wheel for 'pypdf2'
ValueError: Couldn't find a pure Python 3 wheel for 'pypdf4'

I suspect that the pyodide and PyPDF2/4 projects must have different notions of what it means for a package to be 'pure'. Does anyone here know what is meant by pure in each of these contexts?

Ideally I would like to be able to use this library in Pyodide. Is there a current plan to extend support to non pure libraries? If so, what does the time line on this look like and if not, what are the main road blocks preventing this. Thanks for your time.

**Comment by hoodmane** [comment, marked as answer]
Well it's two things: they need to be pure *and* they need to have a wheel. Pypdf doesn't seem to have wheels (made with `python setup.py bdist_wheel`) they only have a source distribution (made with `python setup.py sdist`). You should ask them to upload wheels to pypi. 

I think what you can do though is make your own wheel by cloning the source and making the bdist_wheel, then serve it yourself and download from the server.

**Reply by grimmerk** [reply]
I guess the "pure" definition also includes fhe package dependent packages are needed to be pure?

**Reply by mathewdblewis** [reply]
Thanks for pointing out that it needs to be a wheel. I'll ask them to upload a wheel.

**Reply by mathewdblewis** [reply]
Thanks to you both for the help so far. I built a wheel myself and uploaded it here:
https://github.com/mathewdblewis/PyPDF4Wheel/blob/main/pypdf4-1.27.0-py3-none-any.whl

I tried following this tutorial in the pyodide docs:
https://pyodide.org/en/stable/usage/loading-packages.html#installing-wheels-from-arbitrary-urls

and I came up with this html code:

	<script src="https://cdn.jsdelivr.net/pyodide/v0.18.1/full/pyodide.js"></script>
	<script type="text/javascript">
		async function main() {
			let pyodide = await loadPyodide({
				indexURL: "https://cdn.jsdelivr.net/pyodide/v0.18.1/full/",
			});
			await pyodide.loadPackage("micropip");
			await pyodide.runPythonAsync(`
			import micropip
			micropip.install('https://github.com/mathewdblewis\
			/PyPDF4Wheel/blob/main/pypdf4-1.27.0-py3-none-any.whl')
		`);
		}
		main();
	</script>

However when I run this in my chromium browser on my mac, I get this error message:

    Access to fetch at 'https://github.com/mathewdblewis/PyPDF4Wheel/blob/main/pypdf4-1.27.0-py3-none-any.whl' from origin 'null' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

I've run into this issue before when doing web development, however in this case, the resource I am accessing is a file on github, so its not clear to me how I would avoid the CORS policy. What would any of you recommend as a workaround here? Thanks again for all your time, It's greatly appreciated.

**Reply by ryanking13** [reply]
Try this instead, `raw.githubusercontent.com` allows cross-origin requests.

```
https://raw.githubusercontent.com/mathewdblewis/PyPDF4Wheel/main/pypdf4-1.27.0-py3-none-any.whl
```

**Reply by mathewdblewis** [reply]
@ryanking13, thanks for the suggestion, you have helped me make progress. Currently I have the following code:

	<script src="https://cdn.jsdelivr.net/pyodide/v0.18.1/full/pyodide.js"></script>

	<script type="text/javascript">
		async function main() {
			let pyodide = await loadPyodide({indexURL:"https://cdn.jsdelivr.net/pyodide/v0.18.1/full/"});
			await pyodide.loadPackage("micropip");
			await pyodide.runPythonAsync(`
				import micropip
				micropip.install('https://raw.githubusercontent.com/mathewdblewis/PyPDF4Wheel/main/pypdf4-1.27.0-py3-none-any.whl')
				import PyPDF4
		`);
		}
		main();
	</script>


However, it gives the following error:

    Uncaught (in promise) PythonError: Traceback (most recent call last):
      File "/lib/python3.9/asyncio/futures.py", line 201, in result
        raise self._exception
      File "/lib/python3.9/asyncio/tasks.py", line 256, in __step
        result = coro.send(None)
      File "/lib/python3.9/site-packages/_pyodide/_base.py", line 494, in eval_code_async
        await CodeRunner(
      File "/lib/python3.9/site-packages/_pyodide/_base.py", line 345, in run_async
        coroutine = eval(self.code, globals, locals)
      File "<exec>", line 4, in <module>
    ModuleNotFoundError: No module named 'PyPDF4'

So it seems the CORS problem is resolved (thanks again @ryanking13 ) however I am still not clear on how to use the pyodide library.

**Reply by hoodmane** [reply]
You need an `await`:
`await micropip.install("...")`

**Reply by mathewdblewis** [reply]
Thanks hoodmane, I should have known to use await. I fixed the code but I still get a similar error. Here's the fixed code:

	<script src="https://cdn.jsdelivr.net/pyodide/v0.18.1/full/pyodide.js"></script>
	<script type="text/javascript">
		async function main() {
			let pyodide = await loadPyodide({indexURL:"https://cdn.jsdelivr.net/pyodide/v0.18.1/full/"});
			await pyodide.loadPackage("micropip");
			await pyodide.runPythonAsync(`
				import micropip
				await micropip.install('https://raw.githubusercontent.com/mathewdblewis/PyPDF4Wheel/main/pypdf4-1.27.0-py3-none-any.whl')
				import PyPDF4
		`);
		}
		main();
	</script>

Here's the error:

    Uncaught (in promise) PythonError: Traceback (most recent call last):
      File "/lib/python3.9/asyncio/futures.py", line 201, in result
        raise self._exception
      File "/lib/python3.9/asyncio/tasks.py", line 256, in __step
        result = coro.send(None)
      File "/lib/python3.9/site-packages/_pyodide/_base.py", line 494, in eval_code_async
        await CodeRunner(
      File "/lib/python3.9/site-packages/_pyodide/_base.py", line 347, in run_async
        await coroutine
      File "<exec>", line 4, in <module>
    ModuleNotFoundError: No module named 'PyPDF4'

**Reply by hoodmane** [reply]
Try:
```py
import pypdf
```

**Reply by mathewdblewis** [reply]
@hoodmane , Thanks! It worked! I'm going to try out using the library now. I'll accept this answer and if a new problem arises, I'll ask a new question.
